### PR TITLE
Fixes #854: Add --tag option to push:artifact.

### LINKED ISSUE #854: CLI-700: Add --tag option to `push:artifact`
If added, it will create a build tag rather than build branch.

### PR DESCRIPTION
**Pull request by grasmash**
**Motivation**

Fixes #854: Add --tag option to push:artifact.

**Proposed changes**

Add a `--tag-name=[tag]` option that will create and push a git tag rather than a branch.

**Alternatives considered**

**Testing steps**

1. Follow the [contribution guide](https://github.com/acquia/cli/blob/master/CONTRIBUTING.md#building-and-testing) to set up your development environment.
2. Clear the kernel cache to pick up new and changed commands: `./bin/acli ckc`
3. `cd` into a drupal application directory
4. `../cli/bin/acli push:artifact --destination-git-branch=[branch] --tag-name=[tag] -v`

**Merge requirements**
- [x] _Bug_, _enhancement_, or _breaking change_ label applied
- [ ] Manual testing by a reviewer

### CODE REVIEW

**Comment by anavarre** [comment]
`-tag` doesn't work.

```
test:~/project (main) $ ./aclidev push:artifact -tag pr857

                                   
  The "-t" option does not exist.  
                                   

push:artifact [--dir DIR] [--no-sanitize] [--dry-run] [-dest-git-urls|--destination-git-urls DESTINATION-GIT-URLS] [-dest-git-branch|--destination-git-branch DESTINATION-GIT-BRANCH] [-tag|--tag-name TAG-NAME] [--] [<environmentId>]
```

Thought that might be a typo and tried `--tag`. Doesn't work either.

```
test:~/project (main) $ ./aclidev push:artifact --tag pr857

                                      
  The "--tag" option does not exist.  
                                      

push:artifact [--dir DIR] [--no-sanitize] [--dry-run] [-dest-git-urls|--destination-git-urls DESTINATION-GIT-URLS] [-dest-git-branch|--destination-git-branch DESTINATION-GIT-BRANCH] [-tag|--tag-name TAG-NAME] [--] [<environmentId>]


                                                                                                                        
 [help] You can find Acquia CLI documentation at https://docs.acquia.com/acquia-cli/                                    
```                                                                                          

Seems odd anyway to have `-tag`. I expect either `-t` or `--tag`

 -----

`--tag-name` does seem to work.                                                                                                                  

test:~/project (main) $ ./aclidev push:artifact --tag-name=pr857

In PushArtifactCommand.php line 93:
                                                                                                                                                   
  Pushing code was aborted because your local Git repository has uncommitted changes. Please either commit, reset, or stash your changes via git.  
                                                                                                                                                   

push:artifact [--dir DIR] [--no-sanitize] [--dry-run] [-dest-git-urls|--destination-git-urls DESTINATION-GIT-URLS] [-dest-git-branch|--destination-git-branch DESTINATION-GIT-BRANCH] [-tag|--tag-name TAG-NAME] [--] [<environmentId>]
```

**Comment by grasmash** [comment]
Yeah I decided that using tag name would be better. What do you think?

**Comment by anavarre** [comment]
What I'm saying is I'm still seeing the `-tag` option:

```
acli push:artifact --help | grep tag
  -tag, --tag-name=TAG-NAME                                          If defined, a tag with this name will be created and pushed
```

I'd suggest we either use `-t` or `--tag` along with `--tag-name`. I don't really like `-tag` nor do I think it's common to do.

File: src/Command/Push/PushArtifactCommand.php [hunk 1]
```diff
@@ -441,10 +450,31 @@ protected function determineDestinationGitBranch() {
     if ($env_var = getenv('ACLI_PUSH_ARTIFACT_DESTINATION_GIT_BRANCH')) {
       return $env_var;
     }
+
+    $this->environment = $this->determineEnvironment($this->input, $this->output, FALSE);
     if (strpos($this->environment->vcs->path, 'tags') === 0) {
       throw new AcquiaCliException("You cannot push to an environment that has a git tag deployed to it. Environment {$this->environment->name} has {$this->environment->vcs->path} deployed. Please select a different environment.");
```
Comment (danepowell) [inline src/Command/Push/PushArtifactCommand.php:456]: Is this error message unnecessary now?

**Review by danepowell — APPROVED** [review]

File: src/Command/Push/PushArtifactCommand.php (hunk 1 above)
Comment (grasmash) [reply src/Command/Push/PushArtifactCommand.php:456]: I think it's still necessary. The naming of variables is tricky and maybe a bit misleading. The "destination-git-branch" which is inferred from the environment is actually used as the source branch and the destination branch when a tag is not being used. 

When a tag is being used, destination-git-branch is used as the source branch and the tag is used as the destination git ref.

Comment (grasmash) [reply src/Command/Push/PushArtifactCommand.php:456]: * `acli push:artifact`
    * prompts to choose environment. the branch of the chosen environment will be the `--destination-git-branch`
    * `--destination-git-branch` is used as the source and destination branch
* `acli push:artifact --destination-git-branch=[branch]`
    * `--destination-git-branch` is used as the source and destination branch
* `acli push:artifact --destination-git-branch=[branch] --tag-name=[tag]`
    * `--destination-git-branch` is used as the source and destination branch

Comment (grasmash) [reply src/Command/Push/PushArtifactCommand.php:483]: I've updated this.

**Comment by grasmash** [comment]
Ok, now I'm using the `--destination-git-tag` option with the `-t` shortcut.

Also,
* `acli push:artifact`
    * prompts to choose environment. the branch of the chosen environment will be the `--destination-git-branch`
    * `--destination-git-branch` is used as the source and destination branch
* `acli push:artifact --destination-git-branch=[branch]`
    * `--destination-git-branch` is used as the source and destination branch
* `acli push:artifact --destination-git-branch=[branch] --source-git-branch=[branch]`
* `acli push:artifact --destination-git-tag=[tag]`
    * Will throw exception `'You must also set the --source-git-branch option when setting the --destination-git-tag option.'`
* `acli push:artifact --destination-git-tag=[tag]  --source-git-branch=[branch]`

**Review by anavarre — APPROVED** [review]
Help looks good now. Seeing all the options mentioned.

```
test:~/project (main *+) $ ../aclidev push:artifact --help
Description:
  Build and push a code artifact to a Cloud Platform environment

Usage:
  push:artifact [options] [--] [<environmentId>]
  push:artifact [<environmentAlias>]
  push:artifact myapp.dev
  push:artifact 12345-abcd1234-1111-2222-3333-0e02b2c3d470
  push:artifact --destination-git-branch=main-build
  push:artifact --source-git-branch=main-build --destination-git-tag=1.0.0
  push:artifact --dest-git-url=example@svn-1.prod.hosting.acquia.com:example.git --dest-branch=main-build

Arguments:
  environmentId                                        The Cloud Platform environment ID or alias

Options:
      --dir=DIR                                        The directory containing the Drupal project to be pushed
      --no-sanitize                                    Do not sanitize the build artifact
      --dry-run                                        Do not push changes to Acquia Cloud
      --destination-git-urls=DESTINATION-GIT-URLS      The URL(s) of your git repository to which the artifact branch will be pushed (multiple values allowed)
      --destination-git-branch=DESTINATION-GIT-BRANCH  The destination branch to push the artifact to
  -t, --destination-git-tag=DESTINATION-GIT-TAG        The destination tag to push the artifact to. Using this option requires also using the --source-git-branch option
  -s, --source-git-branch=SOURCE-GIT-BRANCH            The destination branch to push the artifact to
  -h, --help                                           Display help for the given command. When no command is given display help for the list command
  -q, --quiet                                          Do not output any message
  -V, --version                                        Display this application version
      --ansi|--no-ansi                                 Force (or disable --no-ansi) ANSI output
  -n, --no-interaction                                 Do not ask any interactive question
  -v|vv|vvv, --verbose                                 Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug

Help:
  This command builds a sanitized deploy artifact by running composer install, removing sensitive files, and committing vendor directories.
  
  Vendor directories and scaffold files are committed to the build artifact even if they are ignored in the source repository.
  
  To run additional build or sanitization steps (e.g. npm install), add a post-install-cmd script to your composer.json file: https://getcomposer.org/doc/articles/scripts.md#command-events
```

Cannot push to a tag as expected.

```
test:~/project (main *+) $ ../aclidev push:artifact
Using Cloud Application anavarred8

 Choose a Cloud Platform environment [Dev, dev (vcs: pipelines-build-master)]:
  [0] Dev, dev (vcs: pipelines-build-master)
  [1] Stage, test (vcs: tags/2020-10-19)
 > 1


In PushArtifactCommand.php line 501:
                                                                                                                                                              
  You cannot push to an environment that has a git tag deployed to it. Environment test has tags/2020-10-19 deployed. Please select a different environment.  
                                                                                                                                                              

push:artifact [--dir DIR] [--no-sanitize] [--dry-run] [--destination-git-urls DESTINATION-GIT-URLS] [--destination-git-branch DESTINATION-GIT-BRANCH] [-t|--destination-git-tag DESTINATION-GIT-TAG] [-s|--source-git-branch SOURCE-GIT-BRANCH] [--] [<environmentId>]


                                                                                                                        
 [help] You can find Acquia CLI documentation at https://docs.acquia.com/acquia-cli/                                    
                                                                                                                        
        You can submit a support ticket at https://insight.acquia.com/support/tickets/new?product=p:cli                 
        Please re-run the command with the -vvv flag and include the full command output in                             
        your support ticket.                                                                                            
                                                                                                                        
```

Area of improvement for us? When the SSH fingerprint is changed, the command fails and can get customers confused.

```
test:~/project (main) $ ../aclidev push:artifact
Using Cloud Application anavarred8

 Choose a Cloud Platform environment [Dev, dev (vcs: pipelines-build-master)]:
  [0] Dev, dev (vcs: pipelines-build-master)
  [1] Stage, test (vcs: blt)
 > 1

 ! [NOTE] Acquia CLI will:                                                                                              
 !                                                                                                                      
 !        - git clone blt from anavarred8@svn-<REDACTED>.prod.hosting.acquia.com:anavarred8.git                              
 !                                                                                                                      
 !        - Compile the contents of /home/ide/project into an artifact in a temporary directory                         
 !                                                                                                                      
 !        - Copy the artifact files into the checked out copy of blt                                                    
 !                                                                                                                      
 !        - Commit changes and push the blt branch to the following git remote(s):                                      
 !                                                                                                                      
 !          anavarred8@svn-<REDACTED>.prod.hosting.acquia.com:anavarred8.git                                                 

    ⠛ Preparing artifact directory...
        and the repository exists.

In PushArtifactCommand.php line 200:
                                                                                                          
  Failed to clone repository from the Cloud Platform: Cloning into '/tmp/acli-push-artifact'...           
  @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@                                             
  @       WARNING: POSSIBLE DNS SPOOFING DETECTED!          @                                             
  @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@                                             
  The RSA host key for svn-<REDACTED>.prod.hosting.acquia.com has changed,                                     
  and the key for the corresponding IP address <REDACTED>                                             
  is unknown. This could either mean that                                                                 
  DNS SPOOFING is happening or the IP address for the host                                                
  and its host key have changed at the same time.                                                         
  @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@                                             
  @    WARNING: REMOTE HOST IDENTIFICATION HAS CHANGED!     @                                             
  @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@                                             
  IT IS POSSIBLE THAT SOMEONE IS DOING SOMETHING NASTY!                                                   
  Someone could be eavesdropping on you right now (man-in-the-middle attack)!                             
  It is also possible that a host key has just been changed.                                              
  The fingerprint for the RSA key sent by the remote host is                                              
  SHA256:<REDACTED>.                                                     
  Please contact your system administrator.                                                               
  Add correct host key in /home/ide/.ssh/known_hosts to get rid of this message.                          
  Offending RSA key in /home/ide/.ssh/known_hosts:1                                                       
    remove with:                                                                                          
    ssh-keygen -f "/home/ide/.ssh/known_hosts" -R svn-<REDACTED>.prod.hosting.acquia.com                       
  RSA host key for svn-<REDACTED>.prod.hosting.acquia.com has changed and you have requested strict checking.  
  Host key verification failed.                                                                           
  fatal: Could not read from remote repository.                                                           
                                                                                                          
  Please make sure you have the correct access rights                                                     
  and the repository exists.                                                                              
                                                                                                          

push:artifact [--dir DIR] [--no-sanitize] [--dry-run] [--destination-git-urls DESTINATION-GIT-URLS] [--destination-git-branch DESTINATION-GIT-BRANCH] [-t|--destination-git-tag DESTINATION-GIT-TAG] [-s|--source-git-branch SOURCE-GIT-BRANCH] [--] [<environmentId>]


                                                                                                                        
 [help] You can find Acquia CLI documentation at https://docs.acquia.com/acquia-cli/                                    
                                                                                                                        
        You can submit a support ticket at https://insight.acquia.com/support/tickets/new?product=p:cli                 
        Please re-run the command with the -vvv flag and include the full command output in                             
        your support ticket.                                                                                            
                                                                                                                        
```

After fixing the fingerprint, the command worked immediately on my `blt` branch.

```
test:~/project (main +) $ rm .acquia-cli.yml 
test:~/project (main *+) $ ../aclidev push:artifact
Using Cloud Application anavarred8

 Choose a Cloud Platform environment [Dev, dev (vcs: pipelines-build-master)]:
  [0] Dev, dev (vcs: pipelines-build-master)
  [1] Stage, test (vcs: blt)
 > 1

 ! [NOTE] Acquia CLI will:                                                                                              
 !                                                                                                                      
 !        - git clone blt from anavarred8@svn-<REDACTED>.prod.hosting.acquia.com:anavarred8.git                              
 !                                                                                                                      
 !        - Compile the contents of /home/ide/project into an artifact in a temporary directory                         
 !                                                                                                                      
 !        - Copy the artifact files into the checked out copy of blt                                                    
 !                                                                                                                      
 !        - Commit changes and push the blt branch to the following git remote(s):                                      
 !                                                                                                                      
 !          anavarred8@svn-<REDACTED>.prod.hosting.acquia.com:anavarred8.git                                                 

    ✔ Preparing artifact directory
    ✔ Generating build artifact
    ✔ Sanitizing build artifact
    ✔ Committing changes (commit hash: fe43fdh8ce6f0b71275495f95a26c34ef537d17c)
    ✔ Pushing changes to blt branch.
```

**Pull request merged**